{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import extern\n",
    "# import polars as pl\n",
    "# import random as rand\n",
    "\n",
    "roundrobin_species_coverages = \"/mnt/hpccs01/work/microbiome/msingle/rossenzhao/phrogs_v4.1/build_spkgs/roundrobin_species_coverage.tsv\"\n",
    "genome_rep_paths = \"/home/n10927662/lyrebird-benchmarking/genome_rep_paths.csv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 1. Sanity check at 10x coverage\n",
    "sanity_genomes_output = \"/home/n10927662/lyrebird-benchmarking/1_sanity_10x/genome_rep_paths.csv\"\n",
    "\n",
    "genome_to_paths = {}\n",
    "with open(genome_rep_paths) as r:\n",
    "    for line in r:\n",
    "        genome, path = line.strip().split(\"\\t\")\n",
    "        genome_to_paths[genome] = path\n",
    "\n",
    "num_over_20 = []\n",
    "num_10_20 = []\n",
    "num_3 = []\n",
    "with open(sanity_genomes_output, \"w+\") as w:\n",
    "    with open(roundrobin_species_coverages, \"r\") as f:\n",
    "        for line in f:\n",
    "            species, coverage = line.strip().split(\"\\t\")\n",
    "            coverage = int(coverage)\n",
    "            if \"IMGVR\" in species:\n",
    "                continue\n",
    "            if coverage >= 20:\n",
    "                if len(num_over_20) < 3:\n",
    "                    num_over_20.append(species)\n",
    "            elif coverage >= 10:\n",
    "                if len(num_10_20) < 3:\n",
    "                    num_10_20.append(species)\n",
    "            elif coverage > 3:\n",
    "                if len(num_3) < 3:\n",
    "                    num_3.append(species)\n",
    "            if len(num_over_20) >= 3 and len(num_10_20) >= 3 and len(num_3) >= 3:\n",
    "                break\n",
    "\n",
    "    for species in num_over_20:\n",
    "        w.write(f\"{species}\\t{genome_to_paths[species]}\\n\")\n",
    "    for species in num_10_20:\n",
    "        w.write(f\"{species}\\t{genome_to_paths[species]}\\n\")\n",
    "    for species in num_3:\n",
    "        w.write(f\"{species}\\t{genome_to_paths[species]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 2. Known strains toy benchmark using CAMI airways toy abundances\n",
    "toy_genomes_output = \"/home/n10927662/lyrebird-benchmarking/2_known_strains_toy/genome_rep_paths.csv\"\n",
    "\n",
    "genome_to_paths = {}\n",
    "with open(genome_rep_paths) as r:\n",
    "    for line in r:\n",
    "        genome, path = line.strip().split(\"\\t\")\n",
    "        genome_to_paths[genome] = path\n",
    "\n",
    "with open(toy_genomes_output, \"w+\") as w:\n",
    "    with open(roundrobin_species_coverages, \"r\") as f:\n",
    "        for line in f:\n",
    "            species, coverage = line.strip().split(\"\\t\")\n",
    "            coverage = int(coverage)\n",
    "            # if \"IMGVR\" in species:\n",
    "            #     continue\n",
    "            if coverage >= 3:\n",
    "                w.write(f\"{species}\\t{genome_to_paths[species]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_covs = [cov for cov in os.listdir(\"/home/n10927662/lyrebird-benchmarking/2_known_strains_toy/coverage_definitions\") if 'coverage' in cov]\n",
    "for cov in old_covs:\n",
    "    os.remove(f\"/home/n10927662/lyrebird-benchmarking/2_known_strains_toy/coverage_definitions/{cov}\")\n",
    "\n",
    "new_covs = [cov for cov in os.listdir(\"/home/n10927662/lyrebird-benchmarking/2_known_strains_toy/coverage_definitions\") if 'abundance' in cov]\n",
    "\n",
    "i = 0\n",
    "for cov in sorted(new_covs):\n",
    "    os.rename(f\"/home/n10927662/lyrebird-benchmarking/2_known_strains_toy/coverage_definitions/{cov}\", f\"/home/n10927662/lyrebird-benchmarking/2_known_strains_toy/coverage_definitions/coverage_{i}.tsv\")\n",
    "    i+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 3. Known species toy benchmark - get non-representative species\n",
    "toy_genomes_output = \"/home/n10927662/lyrebird-benchmarking/3_known_species_toy/genome_rep_paths.csv\"\n",
    "cluster_info = \"/work/microbiome/msingle/rossenzhao/IMG-VR-4.1/ictv-imgvr-clusters.tsv\"\n",
    "\n",
    "# genome_to_paths = {}\n",
    "# with open(genome_rep_paths) as r:\n",
    "#     for line in r:\n",
    "#         genome, path = line.strip().split(\"\\t\")\n",
    "#         genome_to_paths[genome] = path\n",
    "\n",
    "reps_to_clusters = {}\n",
    "clusters_to_paths = {}\n",
    "\n",
    "with open(cluster_info) as r:\n",
    "    for line in r:\n",
    "        rep_path, cluster_path = line.strip().split(\"\\t\")\n",
    "        rep = rep_path.split('/')[-1].rsplit('.', 1)[0]\n",
    "        cluster = cluster_path.split('/')[-1].rsplit('.', 1)[0]\n",
    "        if rep not in reps_to_clusters:\n",
    "            reps_to_clusters[rep] = [cluster]\n",
    "        else:\n",
    "            reps_to_clusters[rep].append(cluster)\n",
    "        clusters_to_paths[cluster] = cluster_path\n",
    "\n",
    "picked_rep_cluster = {}\n",
    "\n",
    "with open(toy_genomes_output, \"w+\") as w:\n",
    "    with open(roundrobin_species_coverages, \"r\") as f:\n",
    "        for line in f:\n",
    "            species, coverage = line.strip().split(\"\\t\")\n",
    "            coverage = int(coverage)\n",
    "            if coverage >= 3:\n",
    "                if species not in reps_to_clusters:\n",
    "                    continue\n",
    "                if len(reps_to_clusters[species]) == 1:\n",
    "                    continue\n",
    "                w.write(f\"{species}\\t{clusters_to_paths[reps_to_clusters[species][-1]]}\\n\")\n",
    "                # picked_reps_cluster[species] = reps_to_clusters[species][0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'IMGVR_UViG_2124908014_000246'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mreps_to_clusters\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mIMGVR_UViG_2124908014_000246\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'IMGVR_UViG_2124908014_000246'"
     ]
    }
   ],
   "source": [
    "### 5. sanity vs microbial\n",
    "sanity_genomes_output = \"/home/n10927662/lyrebird-benchmarking/5_sanity_microbial/genome_rep_paths.csv\"\n",
    "microbial_filepaths = \"/work/microbiome/db/gtdb/gtdb_release214/bac120_taxonomy_r214.tsv\"\n",
    "\n",
    "genome_to_paths = {}\n",
    "with open(genome_rep_paths) as r:\n",
    "    for line in r:\n",
    "        genome, path = line.strip().split(\"\\t\")\n",
    "        genome_to_paths[genome] = path\n",
    "\n",
    "num_over_20 = []\n",
    "with open(sanity_genomes_output, \"w+\") as w:\n",
    "    with open(roundrobin_species_coverages, \"r\") as f:\n",
    "        for line in f:\n",
    "            species, coverage = line.strip().split(\"\\t\")\n",
    "            coverage = int(coverage)\n",
    "            if \"IMGVR\" in species:\n",
    "                continue\n",
    "            if coverage >= 20:\n",
    "                if len(num_over_20) < 50:\n",
    "                    num_over_20.append(species)\n",
    "            if len(num_over_20) >= 50:\n",
    "                break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
