{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(350470, 12)\n"
     ]
    }
   ],
   "source": [
    "contig_otus = pl.read_csv('/home/n10927662/rossenzhao/land-use-viromes/lyrebird-0.3.1/lyrebird_0.3.1_contigs.tsv', separator='\\t')\n",
    "print(contig_otus.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_quality_mags = set()\n",
    "with open(\"/home/n10927662/rossenzhao/land-use-viromes/hq-complete-ids.txt\") as f:\n",
    "    for line in f:\n",
    "        high_quality_mags.add(line.strip())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## only root and d__Viruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.522361813600188"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### number of genes per contig\n",
    "contig_to_genes = {}\n",
    "for line in contig_otus.iter_rows():\n",
    "    gene = line[0]\n",
    "    read_names = line[6]\n",
    "    taxonomy = line[5]\n",
    "    if taxonomy == \"Root\" or \"d__Viruses\" in taxonomy:\n",
    "        for read_name in read_names.split():\n",
    "            contig = read_name.rsplit('_', 3)[0]\n",
    "            # if contig not in high_quality_mags:\n",
    "            #     continue\n",
    "            if contig not in contig_to_genes:\n",
    "                contig_to_genes[contig] = set()\n",
    "            contig_to_genes[contig].add(gene)\n",
    "\n",
    "avg_genes_per_contig = sum(len(genes) for genes in contig_to_genes.values()) / len(contig_to_genes)\n",
    "avg_genes_per_contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7011014, 6)\n",
      "(628571, 6)\n",
      "(4704250, 6)\n"
     ]
    }
   ],
   "source": [
    "# all_otus = pl.read_csv('/work/microbiome/msingle/rossenzhao/land-use-viromes/lyrebird-0.3.1/otu_tables/*.tsv', separator='\\t')\n",
    "all_otus = pl.read_csv('/work/microbiome/msingle/rossenzhao/land-use-viromes/lyrebird-0.3.1/lyrebird_0.3.1_clustered_otu_table.tsv', separator='\\t')\n",
    "print(all_otus.shape)\n",
    "# drop read_unaligned_sequences and equal_best_hit_taxonomies column\n",
    "# all_otus = all_otus.drop(['read_unaligned_sequences', 'equal_best_hit_taxonomies'])\n",
    "root_otus = all_otus.filter(pl.col('taxonomy') == \"Root\")\n",
    "print(root_otus.shape)\n",
    "viral_otus = all_otus.filter(pl.col('taxonomy').str.contains(\"d__Viruses\"))\n",
    "viral_otus = pl.concat([viral_otus, root_otus])\n",
    "viral_otus.head()\n",
    "del all_otus, root_otus\n",
    "print(viral_otus.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(335359, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5, 17)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>gene</th><th>sample</th><th>sequence</th><th>num_hits</th><th>coverage</th><th>taxonomy</th><th>gene_right</th><th>sample_right</th><th>num_hits_right</th><th>coverage_right</th><th>taxonomy_right</th><th>read_names</th><th>nucleotides_aligned</th><th>taxonomy_by_known?</th><th>read_unaligned_sequences</th><th>equal_best_hit_taxonomies</th><th>taxonomy_assignment_method</th></tr><tr><td>str</td><td>str</td><td>str</td><td>i64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194036_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>23130</td><td>38468.8</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194039_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>8759</td><td>14567.59</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194055_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>12</td><td>19.96</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194046_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>11</td><td>18.29</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194056_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>11</td><td>18.29</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5, 17)\n",
       "┌───────────┬───────────┬───────────┬──────────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ gene      ┆ sample    ┆ sequence  ┆ num_hits ┆ … ┆ taxonomy_ ┆ read_unal ┆ equal_bes ┆ taxonomy_ │\n",
       "│ ---       ┆ ---       ┆ ---       ┆ ---      ┆   ┆ by_known? ┆ igned_seq ┆ t_hit_tax ┆ assignmen │\n",
       "│ str       ┆ str       ┆ str       ┆ i64      ┆   ┆ ---       ┆ uences    ┆ onomies   ┆ t_method  │\n",
       "│           ┆           ┆           ┆          ┆   ┆ bool      ┆ ---       ┆ ---       ┆ ---       │\n",
       "│           ┆           ┆           ┆          ┆   ┆           ┆ str       ┆ str       ┆ str       │\n",
       "╞═══════════╪═══════════╪═══════════╪══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 23130    ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 36_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 8759     ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 39_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 12       ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 55_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 11       ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 46_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 11       ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 56_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "└───────────┴───────────┴───────────┴──────────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join to contigs via contig OTU table\n",
    "contig_otus_join = viral_otus.join(contig_otus, on='sequence', how='inner')\n",
    "print(contig_otus_join.shape)\n",
    "contig_otus_join.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample1_lyrebird_detected = set() # contig sequences detected by Lyrebird from reads\n",
    "# for i in contig_otus_join['read_names_right']:\n",
    "for i in contig_otus_join['read_names']:\n",
    "    for j in i.split(' '):\n",
    "        sample1_lyrebird_detected.add(j.rsplit('_', 3)[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55161\n",
      "92.51165598899809\n"
     ]
    }
   ],
   "source": [
    "# contigs detected from contig otus\n",
    "contig_otus_detected = set() # contig sequences detected by Lyrebird from contigs\n",
    "for i in contig_otus.filter((pl.col('taxonomy').str.contains(\"d__Viruses\")) | (pl.col('taxonomy') == \"Root\"))['read_names']:\n",
    "# for i in contig_otus['read_names']:\n",
    "    for j in i.split(' '):\n",
    "        contig_otus_detected.add(j.rsplit('_', 3)[0])\n",
    "print(len(contig_otus_detected))\n",
    "print(len(contig_otus_detected)/59626 * 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number detected ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47153\n",
      "79.08127327004998\n"
     ]
    }
   ],
   "source": [
    "print(len(sample1_lyrebird_detected))\n",
    "print(len(sample1_lyrebird_detected) / 59626 * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "254370\n"
     ]
    }
   ],
   "source": [
    "### unique contig window sequences\n",
    "viral_contig_windows = contig_otus.filter((pl.col('taxonomy').str.contains(\"d__Viruses\")) | (pl.col('taxonomy') == \"Root\"))\n",
    "unique_contig_windows = set()\n",
    "for sequence in viral_contig_windows[\"sequence\"]:\n",
    "# for sequence in contig_otus_join[\"sequence\"]:\n",
    "    unique_contig_windows.add(sequence)\n",
    "print(len(unique_contig_windows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3473530\n"
     ]
    }
   ],
   "source": [
    "# unique reads window sequences\n",
    "unique_reads_windows = set()\n",
    "# for sequence in viral_otus[\"sequence\"]:\n",
    "for sequence in viral_otus[\"sequence\"]:\n",
    "    unique_reads_windows.add(sequence)\n",
    "print(len(unique_reads_windows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "est_num_contig_windows if lyrebird detected all: 321656.42949547217\n",
      "reads + contigs windows: 131947\n",
      "reads-only windows: 3341583\n",
      "number of lyrebird windows vs contigs: 10.388671556298048\n",
      "number of non-contig viruses: 506260.9341283608\n",
      "total number of unique lyrebird windows: 526251.3433073142\n",
      "8.81552102833212\n"
     ]
    }
   ],
   "source": [
    "num_read_and_contig_windows = len(unique_reads_windows.intersection(unique_contig_windows))\n",
    "num_only_read_windows = len(unique_reads_windows - unique_contig_windows)\n",
    "est_num_contig_windows = len(unique_contig_windows) * 59626 / len(sample1_lyrebird_detected)\n",
    "print(\"est_num_contig_windows if lyrebird detected all: \" + str(est_num_contig_windows)) # estimated number of contig windows if lyrebird detected all contigs\n",
    "print(\"reads + contigs windows: \" + str(num_read_and_contig_windows)) # number of windows that are both in reads and contigs\n",
    "print(\"reads-only windows: \" + str(num_only_read_windows)) # number of windows that are only in reads\n",
    "print(\"number of lyrebird windows vs contigs: \" + str(num_only_read_windows / est_num_contig_windows)) # how many times more windows lyrebird detected vs contigs\n",
    "print(\"number of non-contig viruses: \" + str(num_only_read_windows / avg_genes_per_contig)) # estimated total number of unique viruses not found in contigs\n",
    "# print(num_only_read_windows / 10.48039840176038) # estimated total number of unique viruses not found in contigs\n",
    "lyrebird_unique_estimate = len(unique_reads_windows) / avg_genes_per_contig\n",
    "# lyrebird_unique_estimate = len(unique_reads_windows) / 10.48039840176038\n",
    "print(\"total number of unique lyrebird windows: \" + str(lyrebird_unique_estimate)) # estimated total number of unique viruses found by lyrebird\n",
    "print(lyrebird_unique_estimate/59696)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## include all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8.829737114859135"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### number of genes per contig\n",
    "contig_to_genes = {}\n",
    "for line in contig_otus.iter_rows():\n",
    "    gene = line[0]\n",
    "    read_names = line[6]\n",
    "    taxonomy = line[5]\n",
    "    for read_name in read_names.split():\n",
    "        contig = read_name.rsplit('_', 3)[0]\n",
    "        if contig not in high_quality_mags:\n",
    "            continue\n",
    "        if contig not in contig_to_genes:\n",
    "            contig_to_genes[contig] = set()\n",
    "        contig_to_genes[contig].add(gene)\n",
    "\n",
    "avg_genes_per_contig = sum(len(genes) for genes in contig_to_genes.values()) / len(contig_to_genes)\n",
    "avg_genes_per_contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7011014, 6)\n"
     ]
    }
   ],
   "source": [
    "# all_otus = pl.read_csv('/work/microbiome/msingle/rossenzhao/land-use-viromes/lyrebird-0.3.1/otu_tables/*.tsv', separator='\\t')\n",
    "all_otus = pl.read_csv('/work/microbiome/msingle/rossenzhao/land-use-viromes/lyrebird-0.3.1/lyrebird_0.3.1_clustered_otu_table.tsv', separator='\\t')\n",
    "print(all_otus.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(445814, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5, 17)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>gene</th><th>sample</th><th>sequence</th><th>num_hits</th><th>coverage</th><th>taxonomy</th><th>gene_right</th><th>sample_right</th><th>num_hits_right</th><th>coverage_right</th><th>taxonomy_right</th><th>read_names</th><th>nucleotides_aligned</th><th>taxonomy_by_known?</th><th>read_unaligned_sequences</th><th>equal_best_hit_taxonomies</th><th>taxonomy_assignment_method</th></tr><tr><td>str</td><td>str</td><td>str</td><td>i64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194036_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>23130</td><td>38468.8</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194039_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>8759</td><td>14567.59</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194055_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>12</td><td>19.96</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194046_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>11</td><td>18.29</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr><tr><td>&quot;phrog_5302&quot;</td><td>&quot;SRR14194056_1&quot;</td><td>&quot;ACCGCCGCCCGCAACGCCTGTAGTGCAAAA…</td><td>11</td><td>18.29</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;phrog_5302&quot;</td><td>&quot;contigs2&quot;</td><td>1</td><td>1.0</td><td>&quot;Root; d__Viruses; p__Urovirico…</td><td>&quot;CG12_NODE_119_length_42942_cov…</td><td>&quot;60&quot;</td><td>false</td><td>&quot;CTGACCTCGCCGAAGCCGGGACCGGCTCGC…</td><td>&quot;[&#x27;phrog_5302~IMGVR_UViG_330000…</td><td>&quot;diamond&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5, 17)\n",
       "┌───────────┬───────────┬───────────┬──────────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ gene      ┆ sample    ┆ sequence  ┆ num_hits ┆ … ┆ taxonomy_ ┆ read_unal ┆ equal_bes ┆ taxonomy_ │\n",
       "│ ---       ┆ ---       ┆ ---       ┆ ---      ┆   ┆ by_known? ┆ igned_seq ┆ t_hit_tax ┆ assignmen │\n",
       "│ str       ┆ str       ┆ str       ┆ i64      ┆   ┆ ---       ┆ uences    ┆ onomies   ┆ t_method  │\n",
       "│           ┆           ┆           ┆          ┆   ┆ bool      ┆ ---       ┆ ---       ┆ ---       │\n",
       "│           ┆           ┆           ┆          ┆   ┆           ┆ str       ┆ str       ┆ str       │\n",
       "╞═══════════╪═══════════╪═══════════╪══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 23130    ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 36_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 8759     ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 39_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 12       ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 55_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 11       ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 46_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "│ phrog_530 ┆ SRR141940 ┆ ACCGCCGCC ┆ 11       ┆ … ┆ false     ┆ CTGACCTCG ┆ ['phrog_5 ┆ diamond   │\n",
       "│ 2         ┆ 56_1      ┆ CGCAACGCC ┆          ┆   ┆           ┆ CCGAAGCCG ┆ 302~IMGVR ┆           │\n",
       "│           ┆           ┆ TGTAGTGCA ┆          ┆   ┆           ┆ GGACCGGCT ┆ _UViG_330 ┆           │\n",
       "│           ┆           ┆ AAA…      ┆          ┆   ┆           ┆ CGC…      ┆ 000…      ┆           │\n",
       "└───────────┴───────────┴───────────┴──────────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join to contigs via contig OTU table\n",
    "contig_otus_join = all_otus.join(contig_otus, on='sequence', how='inner')\n",
    "print(contig_otus_join.shape)\n",
    "contig_otus_join.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample1_lyrebird_detected = set() # contig sequences detected by Lyrebird from reads\n",
    "# for i in contig_otus_join['read_names_right']:\n",
    "for i in contig_otus_join['read_names']:\n",
    "    for j in i.split(' '):\n",
    "        sample1_lyrebird_detected.add(j.rsplit('_', 3)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57656\n",
      "96.69607218327575\n"
     ]
    }
   ],
   "source": [
    "# contigs detected from contig otus\n",
    "contig_otus_detected = set() # contig sequences detected by Lyrebird from contigs\n",
    "# for i in contig_otus.filter((pl.col('taxonomy').str.contains(\"d__Viruses\")) | (pl.col('taxonomy') == \"Root\"))['read_names']:\n",
    "for i in contig_otus['read_names']:\n",
    "    for j in i.split(' '):\n",
    "        contig_otus_detected.add(j.rsplit('_', 3)[0])\n",
    "print(len(contig_otus_detected))\n",
    "print(len(contig_otus_detected)/59626 * 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number detected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52016\n",
      "87.23711132727334\n"
     ]
    }
   ],
   "source": [
    "print(len(sample1_lyrebird_detected))\n",
    "print(len(sample1_lyrebird_detected) / 59626 * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "193883\n"
     ]
    }
   ],
   "source": [
    "### unique contig window sequences\n",
    "# viral_contig_windows = contig_otus.filter((pl.col('taxonomy').str.contains(\"d__Viruses\")) | (pl.col('taxonomy') == \"Root\"))\n",
    "unique_contig_windows = set()\n",
    "# for sequence in viral_contig_windows[\"sequence\"]:\n",
    "for sequence in contig_otus_join[\"sequence\"]:\n",
    "    unique_contig_windows.add(sequence)\n",
    "print(len(unique_contig_windows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5230130\n"
     ]
    }
   ],
   "source": [
    "# unique reads window sequences\n",
    "unique_reads_windows = set()\n",
    "for sequence in all_otus[\"sequence\"]:\n",
    "    unique_reads_windows.add(sequence)\n",
    "print(len(unique_reads_windows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "est_num_contig_windows if lyrebird detected all: 222248.30356044293\n",
      "reads + contigs windows: 193883\n",
      "reads-only windows: 5036247\n",
      "number of lyrebird windows vs contigs: 22.660451932900067\n",
      "number of non-contig viruses: 570373.3796926689\n",
      "total number of unique lyrebird windows: 592331.3380642408\n",
      "9.922462779151715\n"
     ]
    }
   ],
   "source": [
    "num_read_and_contig_windows = len(unique_reads_windows.intersection(unique_contig_windows))\n",
    "num_only_read_windows = len(unique_reads_windows - unique_contig_windows)\n",
    "est_num_contig_windows = len(unique_contig_windows) * 59626 / len(sample1_lyrebird_detected)\n",
    "print(\"est_num_contig_windows if lyrebird detected all: \" + str(est_num_contig_windows)) # estimated number of contig windows if lyrebird detected all contigs\n",
    "print(\"reads + contigs windows: \" + str(num_read_and_contig_windows)) # number of windows that are both in reads and contigs\n",
    "print(\"reads-only windows: \" + str(num_only_read_windows)) # number of windows that are only in reads\n",
    "print(\"number of lyrebird windows vs contigs: \" + str(num_only_read_windows / est_num_contig_windows)) # how many times more windows lyrebird detected vs contigs\n",
    "print(\"number of non-contig viruses: \" + str(num_only_read_windows / avg_genes_per_contig)) # estimated total number of unique viruses not found in contigs\n",
    "lyrebird_unique_estimate = len(unique_reads_windows) / avg_genes_per_contig\n",
    "print(\"total number of unique lyrebird windows: \" + str(lyrebird_unique_estimate)) # estimated total number of unique viruses found by lyrebird\n",
    "print(lyrebird_unique_estimate/59696)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## more"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36697875.337290645\n",
      "5559850.121568997\n",
      "111197002431.37994\n"
     ]
    }
   ],
   "source": [
    "covs = 0\n",
    "for cov in viral_otus[\"coverage\"]:\n",
    "    covs += float(cov)\n",
    "\n",
    "print(covs)\n",
    "print(covs/avg_genes_per_contig)\n",
    "print(covs/avg_genes_per_contig * 20000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52820504.7455288\n"
     ]
    }
   ],
   "source": [
    "total_covs = 0\n",
    "for cov in all_otus[\"coverage\"]:\n",
    "    total_covs += float(cov)\n",
    "print(total_covs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "949057697477\n"
     ]
    }
   ],
   "source": [
    "fastq_counts = \"/home/n10927662/rossenzhao/land-use-viromes/fq_count.txt\"\n",
    "basecount=0\n",
    "for line in open(fastq_counts):\n",
    "    bases = int(line.strip().split('\\t')[-1])\n",
    "    basecount += bases\n",
    "print(basecount)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.29291423147135576"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(covs/avg_genes_per_contig * 50000)/basecount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../lyrebird_vs_contigs3.csv\", \"w+\") as f:\n",
    "    f.write(\"category,count\\n\")\n",
    "    f.write(\"OTUs only in reads,\" + str(num_only_read_windows) + \"\\n\")\n",
    "    f.write(\"OTUs in contigs,\" + str(len(unique_contig_windows)) + \"\\n\")\n",
    "\n",
    "with open(\"../contig_with_otu_vs_not3.csv\", \"w+\") as f:\n",
    "    f.write(\"category,count\\n\")\n",
    "    f.write(\"Contigs without OTUs in reads,\" + str(59696-len(sample1_lyrebird_detected)) + \"\\n\")\n",
    "    f.write(\"Contigs with OTUs in reads,\" + str(len(sample1_lyrebird_detected)) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "162\n",
      "37\n"
     ]
    }
   ],
   "source": [
    "phanta_results = \"/home/n10927662/rossenzhao/land-use-viromes/phanta/outputs/final_merged_outputs/relative_taxonomic_abundance.txt\"\n",
    "phanta_viral_count = 0\n",
    "with open(phanta_results) as f:\n",
    "    for line in f:\n",
    "        if \"superkingdom_Viruses\" in line:\n",
    "            phanta_viral_count += 1\n",
    "print(phanta_viral_count)\n",
    "\n",
    "import os\n",
    "metaphlan4_folder = \"/home/n10927662/rossenzhao/land-use-viromes/metaphlan/prof\"\n",
    "metaphlan_viral_count = 0\n",
    "for filename in os.listdir(metaphlan4_folder):\n",
    "    if \"vsc.txt\" in filename:\n",
    "        with open(os.path.join(metaphlan4_folder, filename)) as f:\n",
    "            metaphlan_viral_count += len(f.readlines())-4\n",
    "print(metaphlan_viral_count)\n",
    "\n",
    "with open(\"../tool_results3.csv\", \"w+\") as f:\n",
    "    f.write(\"tool,count\\n\")\n",
    "    f.write(\"lyrebird,\" + str(int(round(lyrebird_unique_estimate))) + \"\\n\")\n",
    "    f.write(\"phanta,\" + str(phanta_viral_count) + \"\\n\")\n",
    "    f.write(\"metaphlan4,\" + str(metaphlan_viral_count) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../genes_per_contig3.csv\", \"w+\") as w:\n",
    "    w.write(\"genes_per_contig\\n\")\n",
    "    for contig in contig_to_genes:\n",
    "        w.write(f\"{len(contig_to_genes[contig])}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
