{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select 500 HQ MAGs and write paths to directory ##\n",
    "- want 100 without proviruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from csv import DictReader\n",
    "from os.path import join\n",
    "\n",
    "quality_report = \"/mnt/hpccs01/work/microbiome/abisko/groopm/60_Emerge_MAGs_v4/Field/Aroney_v3/checkm2/quality_report.tsv\"\n",
    "genome_dir = \"/mnt/hpccs01/work/microbiome/abisko/groopm/60_Emerge_MAGs_v4/Field/Aroney_v3/genomes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9951"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 0\n",
    "with open(\"hq_mag_filepaths.txt\", \"w\") as w:\n",
    "    with open(quality_report, \"r\") as f:\n",
    "        reader = DictReader(f, delimiter=\"\\t\")\n",
    "        for row in reader:\n",
    "            if row[\"Completeness\"] >= \"95\" and row[\"Contamination\"] <= \"5\":\n",
    "                genome_name = row[\"Name\"]\n",
    "                genome_path = join(genome_dir, genome_name + \".fna\")\n",
    "                if os.path.exists(genome_path):\n",
    "                    w.write(genome_name + \"\\t\" + genome_path + \"\\n\")\n",
    "                    i += 1\n",
    "                else:\n",
    "                    print(f\"Warning: {genome_path} does not exist.\")\n",
    "i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# i = 0\n",
    "# j = 0\n",
    "# genomes = []\n",
    "# for folder in os.listdir(\"genomad\"):\n",
    "#     if not os.path.isdir(join(\"genomad\", folder)):\n",
    "#         continue\n",
    "#     provirus_file = join(\"genomad\", folder, f\"{folder}_find_proviruses\", f\"{folder}_provirus.tsv\")\n",
    "#     if not os.path.exists(provirus_file):\n",
    "#         # print(f\"Warning: {provirus_file} does not exist.\")\n",
    "#         continue\n",
    "#     i += 1\n",
    "#     with open(provirus_file, \"r\") as f:\n",
    "#         next(f)\n",
    "#         if len(f.readlines()) > 0:\n",
    "#             j += 1\n",
    "#         else:\n",
    "#             genomes.append(folder)\n",
    "# print(f\"{i} folders processed, with {j} containing proviruses.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9951 folders processed, with 4927 containing proviruses.\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "j = 0\n",
    "genomes = []\n",
    "for folder in os.listdir(\"genomad\"):\n",
    "    if not os.path.isdir(join(\"genomad\", folder)):\n",
    "        continue\n",
    "    genome_file = join(\"genomad\", folder, f\"{folder}_summary\", f\"{folder}_virus_summary.tsv\")\n",
    "    if not os.path.exists(genome_file):\n",
    "        # print(f\"Warning: {genome_file} does not exist.\")\n",
    "        continue\n",
    "    i += 1\n",
    "    with open(genome_file, \"r\") as f:\n",
    "        next(f)\n",
    "        if len(f.readlines()) > 0:\n",
    "            j += 1\n",
    "        else:\n",
    "            genomes.append(folder)\n",
    "print(f\"{i} folders processed, with {j} containing proviruses.\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2520377/930150255.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_df[\"genome_size\"] = filtered_df[\"user_genome\"].map(genome_sizes)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "arc_tax = \"/mnt/hpccs01/work/microbiome/abisko/groopm/60_Emerge_MAGs_v4/Field/Aroney_v3/gtdbtk/classify/gtdbtk.ar53.summary.tsv\"\n",
    "bac_tax = \"/mnt/hpccs01/work/microbiome/abisko/groopm/60_Emerge_MAGs_v4/Field/Aroney_v3/gtdbtk/classify/gtdbtk.bac120.summary.tsv\"\n",
    "\n",
    "headers = [\"user_genome\", \"classification\", \"genome_size\"]\n",
    "# read quality report to get genome sizes\n",
    "with open(quality_report, \"r\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    genome_sizes = {row[\"Name\"]: row[\"Genome_Size\"] for row in reader}\n",
    "\n",
    "# read arc taxonomy\n",
    "arc_df = pd.read_csv(arc_tax, sep=\"\\t\", usecols=[\"user_genome\", \"classification\"])\n",
    "# read bac taxonomy\n",
    "bac_df = pd.read_csv(bac_tax, sep=\"\\t\", usecols=[\"user_genome\", \"classification\"])\n",
    "# merge both dataframes\n",
    "merged_df = pd.concat([arc_df, bac_df], ignore_index=True)\n",
    "# filter for genomes without proviruses\n",
    "filtered_df = merged_df[merged_df[\"user_genome\"].isin(genomes)]\n",
    "# add genome sizes\n",
    "filtered_df[\"genome_size\"] = filtered_df[\"user_genome\"].map(genome_sizes)\n",
    "# save to file\n",
    "filtered_df.to_csv(\"genomes_without_proviruses_taxonomy.tsv\", sep=\"\\t\", index=False, columns=headers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "user_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "classification",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "genome_size",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "8de62e19-530b-4674-b811-a91dea95966c",
       "rows": [
        [
         "346",
         "binchicken_emerge_co106.403",
         "d__Archaea;p__Thermoproteota;c__Bathyarchaeia;o__Bathyarchaeales;f__UBA233;g__;s__",
         "1470543"
        ],
        [
         "658",
         "binchicken_emerge_co185.459",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2483117"
        ],
        [
         "742",
         "binchicken_emerge_co217.349",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2707623"
        ],
        [
         "1355",
         "binchicken_emerge_co82.78",
         "d__Archaea;p__Nanoarchaeota;c__Nanoarchaeia;o__SCGC-AAA011-G17;f__VGLX01;g__;s__",
         "976003"
        ],
        [
         "1843",
         "binchicken_fen_co180.296",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "1984192"
        ],
        [
         "1864",
         "binchicken_fen_co181.503",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2113438"
        ],
        [
         "1961",
         "binchicken_fen_co193.250",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2044008"
        ],
        [
         "1977",
         "binchicken_fen_co195.274",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2126488"
        ],
        [
         "2000",
         "binchicken_fen_co197.448",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2657083"
        ],
        [
         "2036",
         "binchicken_fen_co202.374",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2160251"
        ],
        [
         "2059",
         "binchicken_fen_co205.432",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2586506"
        ],
        [
         "2204",
         "binchicken_fen_co226.403",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "2093863"
        ],
        [
         "2258",
         "binchicken_fen_co234.76",
         "d__Archaea;p__Nanoarchaeota;c__Nanoarchaeia;o__Woesearchaeales;f__UBA10107;g__;s__",
         "1055794"
        ],
        [
         "2315",
         "binchicken_palsa_co172.327",
         "d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__UBA472;g__;s__",
         "1965117"
        ],
        [
         "3165",
         "binchicken_bog_co210.562",
         "d__Bacteria;p__Eremiobacterota;c__Eremiobacteria;o__Baltobacterales;f__Baltobacteraceae;g__;s__",
         "3215458"
        ],
        [
         "3583",
         "binchicken_bog_co214.331",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "1048354"
        ],
        [
         "3618",
         "binchicken_bog_co214.391",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "775528"
        ],
        [
         "3846",
         "binchicken_bog_co216.174",
         "d__Bacteria;p__Acidobacteriota;c__Terriglobia;o__Terriglobales;f__Acidobacteriaceae;g__;s__",
         "5671060"
        ],
        [
         "4224",
         "binchicken_bog_co218.38",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "5825848"
        ],
        [
         "4425",
         "binchicken_bog_co219.429",
         "d__Bacteria;p__Acidobacteriota;c__Terriglobia;o__Terriglobales;f__Koribacteraceae;g__;s__",
         "3341518"
        ],
        [
         "4537",
         "binchicken_bog_co220.432",
         "d__Bacteria;p__Patescibacteria;c__Doudnabacteria;o__UBA920;f__;g__;s__",
         "872438"
        ],
        [
         "4555",
         "binchicken_bog_co220.96",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "787497"
        ],
        [
         "4935",
         "binchicken_bog_co225.134",
         "d__Bacteria;p__FCPU426;c__Palsa-1180;o__Palsa-1180;f__Palsa-1180;g__;s__",
         "3886009"
        ],
        [
         "5098",
         "binchicken_bog_co226.333",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "868031"
        ],
        [
         "5194",
         "binchicken_bog_co227.63",
         "d__Bacteria;p__FCPU426;c__Palsa-1180;o__Palsa-1180;f__Palsa-1180;g__;s__",
         "4052593"
        ],
        [
         "5205",
         "binchicken_bog_co228.131",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "823691"
        ],
        [
         "5291",
         "binchicken_bog_co228.439",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "5807312"
        ],
        [
         "6103",
         "binchicken_bog_co236.209",
         "d__Bacteria;p__FCPU426;c__Palsa-1180;o__Palsa-1180;f__Palsa-1180;g__;s__",
         "3523782"
        ],
        [
         "6110",
         "binchicken_bog_co236.24",
         "d__Bacteria;p__FCPU426;c__Palsa-1180;o__Palsa-1180;f__Palsa-1180;g__;s__",
         "3405990"
        ],
        [
         "6312",
         "binchicken_bog_co238.353",
         "d__Bacteria;p__JAJYCY01;c__JAJYCY01;o__JAJYGY01;f__JAJYGY01;g__;s__",
         "2656146"
        ],
        [
         "6473",
         "binchicken_bog_co240.219",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "5948360"
        ],
        [
         "6538",
         "binchicken_bog_co240.370",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "783411"
        ],
        [
         "6780",
         "binchicken_bog_co242.658",
         "d__Bacteria;p__Patescibacteria;c__Paceibacteria;o__UBA9973;f__PALSA-1337;g__;s__",
         "789588"
        ],
        [
         "7004",
         "binchicken_bog_co245.220",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "851403"
        ],
        [
         "7116",
         "binchicken_bog_co245.568",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "5472157"
        ],
        [
         "7839",
         "binchicken_bog_co253.340",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "712993"
        ],
        [
         "8526",
         "binchicken_bog_co260.281",
         "d__Bacteria;p__Acidobacteriota;c__Terriglobia;o__Terriglobales;f__Koribacteraceae;g__;s__",
         "3237935"
        ],
        [
         "8636",
         "binchicken_bog_co262.182",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "958708"
        ],
        [
         "9003",
         "binchicken_bog_co266.335",
         "d__Bacteria;p__Acidobacteriota;c__Terriglobia;o__Terriglobales;f__Koribacteraceae;g__;s__",
         "3572532"
        ],
        [
         "9027",
         "binchicken_bog_co266.7",
         "d__Bacteria;p__JAJYCY01;c__JAJYCY01;o__JAJYGY01;f__JAJYGY01;g__;s__",
         "2561250"
        ],
        [
         "9315",
         "binchicken_bog_co269.90",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "994579"
        ],
        [
         "9756",
         "binchicken_bog_co274.98",
         "d__Bacteria;p__FCPU426;c__Palsa-1180;o__Palsa-1180;f__Palsa-1180;g__;s__",
         "3966128"
        ],
        [
         "9944",
         "binchicken_bog_co276.49",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "831146"
        ],
        [
         "9989",
         "binchicken_bog_co277.27",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "5976626"
        ],
        [
         "10010",
         "binchicken_bog_co277.335",
         "d__Bacteria;p__Acidobacteriota;c__Terriglobia;o__Terriglobales;f__Koribacteraceae;g__;s__",
         "4439584"
        ],
        [
         "10189",
         "binchicken_bog_co278.532",
         "d__Bacteria;p__Patescibacteria;c__Saccharimonadia;o__Saccharimonadales;f__UBA4665;g__;s__",
         "828965"
        ],
        [
         "10289",
         "binchicken_bog_co279.61",
         "d__Bacteria;p__JAJYCY01;c__JAJYCY01;o__JAJYGY01;f__JAJYGY01;g__;s__",
         "2295271"
        ],
        [
         "10393",
         "binchicken_bog_co281.194",
         "d__Bacteria;p__JAJYCY01;c__JAJYCY01;o__JAJYGY01;f__JAJYGY01;g__;s__",
         "2664400"
        ],
        [
         "10414",
         "binchicken_bog_co282.15",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "5981514"
        ],
        [
         "10491",
         "binchicken_bog_co283.162",
         "d__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Streptosporangiales;f__Streptosporangiaceae;g__;s__",
         "6021732"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 449
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>user_genome</th>\n",
       "      <th>classification</th>\n",
       "      <th>genome_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>binchicken_emerge_co106.403</td>\n",
       "      <td>d__Archaea;p__Thermoproteota;c__Bathyarchaeia;...</td>\n",
       "      <td>1470543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>binchicken_emerge_co185.459</td>\n",
       "      <td>d__Archaea;p__Thermoplasmatota;c__Thermoplasma...</td>\n",
       "      <td>2483117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>binchicken_emerge_co217.349</td>\n",
       "      <td>d__Archaea;p__Thermoplasmatota;c__Thermoplasma...</td>\n",
       "      <td>2707623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>binchicken_emerge_co82.78</td>\n",
       "      <td>d__Archaea;p__Nanoarchaeota;c__Nanoarchaeia;o_...</td>\n",
       "      <td>976003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1843</th>\n",
       "      <td>binchicken_fen_co180.296</td>\n",
       "      <td>d__Archaea;p__Thermoplasmatota;c__Thermoplasma...</td>\n",
       "      <td>1984192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57941</th>\n",
       "      <td>binchicken_palsa_co192.599</td>\n",
       "      <td>d__Bacteria;p__Eremiobacterota;c__Eremiobacter...</td>\n",
       "      <td>2986257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58180</th>\n",
       "      <td>binchicken_palsa_co198.373</td>\n",
       "      <td>d__Bacteria;p__Myxococcota;c__Myxococcia;o__My...</td>\n",
       "      <td>5126854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58267</th>\n",
       "      <td>binchicken_palsa_co199.190</td>\n",
       "      <td>d__Bacteria;p__Verrucomicrobiota;c__Verrucomic...</td>\n",
       "      <td>3554215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58272</th>\n",
       "      <td>binchicken_palsa_co199.21</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>4138486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58375</th>\n",
       "      <td>binchicken_palsa_co199.62</td>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Acidimicrobii...</td>\n",
       "      <td>4425697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>449 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       user_genome  \\\n",
       "346    binchicken_emerge_co106.403   \n",
       "658    binchicken_emerge_co185.459   \n",
       "742    binchicken_emerge_co217.349   \n",
       "1355     binchicken_emerge_co82.78   \n",
       "1843      binchicken_fen_co180.296   \n",
       "...                            ...   \n",
       "57941   binchicken_palsa_co192.599   \n",
       "58180   binchicken_palsa_co198.373   \n",
       "58267   binchicken_palsa_co199.190   \n",
       "58272    binchicken_palsa_co199.21   \n",
       "58375    binchicken_palsa_co199.62   \n",
       "\n",
       "                                          classification genome_size  \n",
       "346    d__Archaea;p__Thermoproteota;c__Bathyarchaeia;...     1470543  \n",
       "658    d__Archaea;p__Thermoplasmatota;c__Thermoplasma...     2483117  \n",
       "742    d__Archaea;p__Thermoplasmatota;c__Thermoplasma...     2707623  \n",
       "1355   d__Archaea;p__Nanoarchaeota;c__Nanoarchaeia;o_...      976003  \n",
       "1843   d__Archaea;p__Thermoplasmatota;c__Thermoplasma...     1984192  \n",
       "...                                                  ...         ...  \n",
       "57941  d__Bacteria;p__Eremiobacterota;c__Eremiobacter...     2986257  \n",
       "58180  d__Bacteria;p__Myxococcota;c__Myxococcia;o__My...     5126854  \n",
       "58267  d__Bacteria;p__Verrucomicrobiota;c__Verrucomic...     3554215  \n",
       "58272  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...     4138486  \n",
       "58375  d__Bacteria;p__Actinomycetota;c__Acidimicrobii...     4425697  \n",
       "\n",
       "[449 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select novel aka species level classification ends with just \"s__\"\n",
    "novel_genomes = filtered_df[filtered_df[\"classification\"].str.endswith(\"g__;s__\")]\n",
    "# novel_genomes = filtered_df[filtered_df[\"classification\"].str.endswith(\"s__\")]\n",
    "novel_genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"novel_genomes_without_proviruses.txt\", 'w') as w:\n",
    "    # for genome in genomes:\n",
    "    #     w.write(genome + \"\\t\" + join(genome_dir, genome + \".fna\") + \"\\n\")\n",
    "    for _, row in novel_genomes.iterrows():\n",
    "        genome_name = row[\"user_genome\"]\n",
    "        genome_path = join(genome_dir, genome_name + \".fna\")\n",
    "        if os.path.exists(genome_path):\n",
    "            w.write(genome_name + \"\\t\" + genome_path + \"\\n\")\n",
    "        else:\n",
    "            print(f\"Warning: {genome_path} does not exist.\")\n",
    "\n",
    "with open(\"novel_genomes_without_proviruses_taxonomy.tsv\", 'w') as w:\n",
    "    novel_genomes.to_csv(w, sep=\"\\t\", index=False, columns=headers)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
