{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "rep_paths = \"/mnt/hpccs01/work/microbiome/db/gtdb/gtdb_release214/genomic_files_reps/gtdb_genomes_reps_r214/out\"\n",
    "basepath = \"/work/microbiome/db/gtdb/gtdb_release214/genomic_files_reps/gtdb_genomes_reps_r214\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(rep_paths) as r:\n",
    "    with open(\"/home/n10927662/rossenzhao/lyrebird-benchmarking/5_sanity_microbial/microbial_rep_paths.csv\", \"w+\") as w:\n",
    "        for line in r:\n",
    "            id = line.split(\"/\")[-1].rsplit(\"_\", 1)[0]\n",
    "            filepath = basepath + line[1:]\n",
    "            # print(f\"{id}\\t{filepath}\")\n",
    "            w.write(f\"{id}\\t{filepath}\")\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "provirus_ids = \"/mnt/hpccs01/home/n10927662/misc_scripts/create_metapackage/gtdb_provirus_ids_r214.tsv\"\n",
    "microbial_rep_paths = \"/home/n10927662/rossenzhao/lyrebird-benchmarking/5_sanity_microbial/microbial_rep_paths.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proviruses = set()\n",
    "with open(provirus_ids) as r:\n",
    "    for line in r:\n",
    "        proviruses.add(line.strip())\n",
    "\n",
    "microbes_without_proviruses = []\n",
    "\n",
    "with open(microbial_rep_paths) as r:\n",
    "    for line in tqdm(r):\n",
    "        id, filepath = line.strip().split(\"\\t\")\n",
    "        for record in SeqIO.parse(filepath, \"fasta\"):\n",
    "            if record.id in proviruses:\n",
    "                break\n",
    "        else:\n",
    "            microbes_without_proviruses.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(microbes_without_proviruses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/home/n10927662/rossenzhao/lyrebird-benchmarking/5_sanity_microbial/microbial_rep_paths_no_proviruses.csv\", \"w+\") as w:\n",
    "    for line in microbes_without_proviruses:\n",
    "        w.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from os.path import join\n",
    "### generate viral rep paths\n",
    "roundrobin_covs = \"/mnt/hpccs01/work/microbiome/msingle/rossenzhao/phrogs_v4.1/0.3/roundrobin_species_coverage.tsv\"\n",
    "genome_dir = \"/mnt/hpccs01/work/microbiome/msingle/rossenzhao/ictv-imgvr/genomes/\"\n",
    "\n",
    "with open(roundrobin_covs) as r:\n",
    "    with open(\"viral_rep_paths.tsv\", \"w+\") as w:\n",
    "        for line in r:\n",
    "            genome_id, cov = line.strip().split(\"\\t\")\n",
    "            if genome_id.startswith(\"NC_\"):\n",
    "                if int(cov) >= 10:\n",
    "                    filepath = join(genome_dir, genome_id + \".fna\")\n",
    "                    if os.path.exists(filepath):\n",
    "                        w.write(f\"{genome_id}\\t{filepath}\\n\")\n",
    "                    else:\n",
    "                        print(f\"File not found: {filepath}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
